gapis/resolve/dependencygraph2/graph_visualization/graph_structure.go

```diff
@@ -0,0 +1,291 @@
+
+	for idNeighbour := range currentNode.outNeighbourIdToEdgeId {
+		neighbour := g.nodeIdToNode[idNeighbour]
+		if (*visitTime)[neighbour.id] == NO_VISITED {
+			g.traverseGraph(neighbour, visitTime, minVisitTime, idInStronglyConnectedComponents, visitedNodesId, currentId, currentTime)
+		}
+		if (*visitTime)[neighbour.id] != VISITED_AND_USED {
+			if (*minVisitTime)[neighbour.id] < (*minVisitTime)[currentNode.id] {
+				(*minVisitTime)[currentNode.id] = (*minVisitTime)[neighbour.id]
+			}
+		}
+	}
+
+	if (*minVisitTime)[currentNode.id] == (*visitTime)[currentNode.id] {
+		for {
+			lastNodeId := (*visitedNodesId)[len(*visitedNodesId)-1]
+			(*visitTime)[lastNodeId] = VISITED_AND_USED
+			*visitedNodesId = (*visitedNodesId)[:len(*visitedNodesId)-1]
+			(*idInStronglyConnectedComponents)[lastNodeId] = *currentId
+			if lastNodeId == currentNode.id {
+				break
+			}
+		}
+		(*currentId)++
+	}
+}
+
+func (g *graph) getIdInStronglyConnectedComponents() []int {
+	currentId := 0
+	currentTime := 1
+	visitTime := make([]int, g.maxNodeId+1)
+	minVisitTime := make([]int, g.maxNodeId+1)
+	idInStronglyConnectedComponents := make([]int, g.maxNodeId+1)
+	visitedNodesId := make([]int, 0)
+
+	for _, currentNode := range g.nodeIdToNode {
+		if visitTime[currentNode.id] == NO_VISITED {
+			g.traverseGraph(currentNode, &visitTime, &minVisitTime, &idInStronglyConnectedComponents, &visitedNodesId, &currentId, &currentTime)
+		}
+	}
+	return idInStronglyConnectedComponents
+}
+
+func (g *graph) makeStronglyConnectedComponentsByCommandTypeId() {
+	newGraph := createGraph(0)
+	for _, currentNode := range g.nodeIdToNode {
+		newNode := g.getNewNode(currentNode.commandTypeId, "")
+		newGraph.addNode(newNode)
+	}
+
+	for _, currentNode := range g.nodeIdToNode {
+		for idNeighbour := range currentNode.outNeighbourIdToEdgeId {
+			neighbour := g.nodeIdToNode[idNeighbour]
+			newGraph.addEdgeBetweenNodesById(currentNode.commandTypeId, neighbour.commandTypeId)
+		}
+	}
+	idInStronglyConnectedComponents := newGraph.getIdInStronglyConnectedComponents()
+	for _, currentNode := range g.nodeIdToNode {
+		id := idInStronglyConnectedComponents[currentNode.commandTypeId]
+		currentNode.label = currentNode.label + "/" + fmt.Sprintf("SCC%d", id)
+	}
+}
+
+func (g *graph) getEdgesInDotFormat() string {
+	output := ""
+	for _, currentEdge := range g.edgeIdToEdge {
+		lines := strconv.Itoa(currentEdge.source.id) + " -> " + strconv.Itoa(currentEdge.sink.id) + ";\n"
+		output += lines
+	}
+	return output
+}
+
+func (g *graph) getNodesInDotFormat() string {
+	output := ""
+	for _, currentNode := range g.nodeIdToNode {
+		lines := strconv.Itoa(currentNode.id) + "[label=" + currentNode.label + "]" + ";\n"
+		output += lines
+	}
+	return output
+}
+
+func (g *graph) getGraphInDotFormat() string {
+	output := "digraph g {\n"
+	output += g.getNodesInDotFormat()
+	output += g.getEdgesInDotFormat()
+	output += "}\n"
+	return output
+}
+
+func (g *graph) getGraphInPbtxtFormat() string {
```

Comment:
This looks like it is trying to write JSON:

I may suggest creating a golang object that matches the json object you are trying to write, and using the json package to output the Golang object directly.

Reply:
This isn't quite JSON. It's TensorFlow's particular flavor of protobuf text format (which is different than the protobuf text format that Go's protobuf package produces).

Reply:
Ahh I apologize, also yay new formats for everyone to write!